pkg/azuredisk/azure_controller_common.go

```diff
@@ -186,25 +184,35 @@ func (c *controllerCommon) AttachDisk(ctx context.Context, diskName, diskURI str
 	}
 	node := strings.ToLower(string(nodeName))
 	diskuri := strings.ToLower(diskURI)
-
-	isMaxDataDiskCountExceeded := c.isMaxDataDiskCountExceeded(ctx, node)
-	if isMaxDataDiskCountExceeded {
-		c.lockMap.LockEntry(node)
-		defer c.lockMap.UnlockEntry(node)
-	}
-
 	requestNum, err := c.insertAttachDiskRequest(diskuri, node, &options)
 	if err != nil {
 		return -1, err
 	}
-	if !isMaxDataDiskCountExceeded {
-		c.lockMap.LockEntry(node)
-		defer c.lockMap.UnlockEntry(node)
 
-		if c.AttachDetachInitialDelayInMs > 0 && requestNum == 1 {
-			klog.V(2).Infof("wait %dms for more requests on node %s, current disk attach: %s", c.AttachDetachInitialDelayInMs, node, diskURI)
-			time.Sleep(time.Duration(c.AttachDetachInitialDelayInMs) * time.Millisecond)
+	var waitForDetach bool
```

Comment:
Considering that with this approach we are serializing the attach calls instead of batching them in case there are detach requests pending, this could:
1. Slow down the attach requests even further as we are serializing the requests, even though there might be available slots to attach multiple disks at the same time.
2. Lead to even more throttling from the CRP side as the number of attach requests will increase.

Instead of this, could we just keep track of the number of the the number of the disks already attached to the node and adjust the batch size accordingly befor making a call to the CRP? That would still allow us to batch and wouldn't aggravate the throttling problem.

Reply:
this PR already reverts the original disk attach serializing fix, and make disk detach happen first, that could help avoiding hitting the `MaximumDataDisksExceeded` error, and even when this happens in rare cases, this PR could also make sure the next disk detach would happen before disk detach happen in retry.

And your suggestion would also make sense to mitigate this issue in another aspect. The tricky part of such fix is that after you get `MaximumDataDisksExceeded` error, current `AttachDisk` call is done with error, need to tell the next `AttachDisk` call to adjust the batch size, while next operation could be already recovered. I will check whether the above PR would fix the issue first, thanks.

Reply:
Do we need to wait for the MaximumDataDisksExceeded  error to be thrown to use this fix? Can we not add this check for every attach call where the number of attach requests > 1?

Reply:
this PR is to avoid MaximumDataDisksExceeded error since it makes sure detach disk would happen first when there are concurrent disk attach & detach requests. Even without this PR, disk attach could be still behind disk detach since there is a node level lock there, this PR just enforces the priority, it's like a priority lock.